### ISSUE TITLE
[Question] Learning why derived setter is not causing suspend (add docs that only direct atom's suspend on write)

### ISSUE THREAD

**Issue by Noitidart** [pmndrs/jotai#276]
Doing set from the dervied atom on an async child atom, is not causing the "Loading..." fallback of the `<Suspend>` to show. 

```
  const [enabled, toggleEnabled] = useAtom(toggleEnabledAtom);
toggleEnable();
```

However doing it on the async child atom itself does show the fallback "Loading...":

```
  const [enabled, setEnabled] = useAtom(enabledAtom);
setEnabled(true);
```

Below is playground. Can you please help me understand this?

Playground - https://codesandbox.io/s/jotai-playground-uuyd5?file=/src/App.js:0-1342
```
import { atom, Provider, useAtom } from "jotai";
import React, { Suspense } from "react";

const enabledAtom = atom("init", async (_get, set, nextValue) => {
  await new Promise((resolve) => setTimeout(resolve, 1000));
  set(enabledAtom, nextValue);
});
const toggleEnabledAtom = atom(
  (get) => get(enabledAtom),
  async (get, set) => {
    const enabled = get(enabledAtom);
    if (enabled) {
      await set(enabledAtom, false);
    } else {
      await set(enabledAtom, true);
    }
  }
);

function Atomed() {
  const [enabled, toggleEnabled] = useAtom(toggleEnabledAtom);
  return (
    <>
      <div>{String(enabled)}</div>
      <div>
        <button
          type="button"
          onClick={() => {
            toggleEnabled();
          }}
        >
          {enabled ? "Disable" : "Enable"}
        </button>
      </div>
    </>
  );
}
export default function App() {
  return (
    <Provider>
      <div className="App">
        <Suspense fallback="Loading...">
          <Atomed />
        </Suspense>
      </div>
    </Provider>
  );
}

```

**Comment by dai-shi** [comment]
Oh, you reached there.
This is something I wanted to discuss with someone.
What you observed is designed behavior, which I knew it could be confusing.
The thing is, we as a React community don't yet know how to deal with `<Suspense>`.
Async `read` is fairly straightforward, and the current behavior is expected.
Async `write` the one you showed is tricky.
I'm not even sure if we should suspend atoms for `write`.
(AFAIK, recoiljs doesn't do this.)

But, using `<Suspense>` for loading state is handy, so I made writing atoms to suspend.
It only works if you `useAtom` the writing atom that has async `write` function.
I think it's technically impossible to suspend dependents atoms on write.
(I actually tried it in my former lib `use-atom`, tracking all previous write dependencies, and it was a disaster. Furthermore, it does not work for the first time, far from optimal.)

In summary, I think there are only two options (unless someone hints a new idea):
1. keep the current behavior, which only works for direct atoms (which can confuse people)
2. drop suspending for async write completely (like recoiljs)

**Comment by Noitidart** [comment]
>Oh, you reached there.

Haha really appreciate your interactions and comments. Going to continue reading the message. Just want to share as I read the first sentence haha.

**Comment by Noitidart** [comment]
Oh thanks for that detailed response. It makes sense we don't yet know how `<Suspense>` will freeze in prod. I think it's a cool design choice to support at on direct atoms. I think I'll try to just document this. The docs might need some work but I know you're busy with the implementation which is way deep. I'm also busy (one man army on an app in prod) ll try to start some tickets to these into the docs.

Side note: "direct atom" is also a nice phrase, it was something I was struggling to reach for and called it "child atom", but it's not necessarily a "child". I think I'll try to put this into a lingo thing for the docs, like "derived atom".

**Comment by dai-shi** [comment]
> I think I'll try to just document this.

That would be wonderful. We need someone also to help organizing docs. See https://github.com/pmndrs/jotai/issues/85#issuecomment-738643270.

**Comment by Noitidart** [comment]
Oh awesome I'll join his efforts!

**Comment by angusryer** [comment]
Thank you for this wonderful, simple library!
I think the issue I'm experiencing is a version of the above problem. Forgive me as I transition into typescript!

I'm defining my atoms in a separate file from the components they're used in, and accessing functions in yet another file that talk to a firestore db. When the user clicks a button to load a saved game, the atom is updated and the useEffect fires, which calls the write-only atom `loadSavedGame`. Inside of that I'm trying to made an async fetch of data from the database to set a number of atoms to.

I have attempted to make the atom itself async and have wrapped my app in <Suspense> (inside the Provider), **but it is not causing the `Suspend` fallback to show.** 

Any thoughts on what I'm mixing up?

``` typescript
// Game.tsx (view component)
export default function Game() {
        const [isLoadingGameOfType, setIsLoadingGameOfType] = useAtom(isLoadingGameOfTypeAtom);
	const [, loadSavedGame] = useAtom(loadSavedGameAction);

        const loadSavedGameRef = useRef(loadSavedGame);
	useEffect(() => {
		if (isLoadingGameOfType === LoadType.SAVED) loadSavedGameRef.current();
	}, [isLoadingGameOfType]);

        return ....

}
```

``` typescript
// gameActions.ts (all atoms are defined here)
export const loadSavedGameAction = atom(null, (get, set) => {
	const playerData = get(playerDataAtom);
	const getAndLoadSavedGame = async () => {
		if (playerData) {
			const data = await getPlayerData(playerData.playerData.playerId); // defined below
			console.log("LOAD SAVED GAME ACTION ==> ", data) // fix: 'data' is always undefined
			if (data) {
				const { lastGameState } = playerData;
				set(gameIdAtom, lastGameState.gameId);
				set(gameStartTimeAtom, lastGameState.gameStartTime);
				set(playerLocationAtom, lastGameState.playerLocation);
				set(playerEnergyAtom, lastGameState.playerEnergy);
				set(playerScoreAtom, lastGameState.playerScore);
				set(playerItemsAtom, lastGameState.playerItems);
			}
		}
	};
	getAndLoadSavedGame();
});
```

``` typescript
// firebaseLogic.ts (all db logic is defined here)
export const getPlayerData = async (
	playerId: string
): Promise<PlayerData | null> => {
	const data = await (await db.collection("users").doc(playerId).get()).data();
	console.log("GET PLAYER DATA ==> ", data?.lastGameState.playerEnergy)
	if (data) return data.playerData;
	return null;
};
```

**Comment by dai-shi** [comment]
> Thank you for this wonderful, simple library!

😍 

> I have attempted to make the write-only atom _itself_ async and have wrapped my app in (inside the Provider), **but it is not causing the `Suspend` fallback to show.**

So, you want to show the suspense fallback on async write?
This is still a new area, and still exploring patterns.
Currently, if you use the atom with async write directly, it will suspend, not the atoms holding values.

Your first attempt sounds right. You would want to provide a _small_ repro (codesandbox) for us to investigate further.

> The code below shows a modification I did based on [#269](https://github.com/pmndrs/jotai/issues/269#issue-790902244), but this is not working either.

To clarify, does it work fine if we don't care suspense fallback?

**Comment by angusryer** [comment]
> So, you want to show the suspense fallback on async write?

By "write", do you mean the calls to `set()` in the write-only atom? If so, then yes: while those atoms haven't been set with the data I'm trying to fetch, I'd like to show the suspense fallback. Is this what you mean by async write?

But none-the-less I got it to work! I realized after looking over the above example from @Noitidart that my use case is very similar, so providing `Suspense` and making the second arg of the write-only derived atom async is what made it work for me. It wouldn't seem to work without providing `Suspense`, to clarify on your last question.

Here's a code sandbox showing the solution: [async fetch to set primitive atoms via a write-only atom in jotai](https://codesandbox.io/s/async-fetch-to-set-primitive-atoms-via-a-write-only-atom-in-jotai-l1qdi?file=/src/App.tsx)

**Comment by dai-shi** [comment]
> By "write", do you mean the calls to `set()` in the write-only atom?

`write` here means the second argument in `atom()`.
https://github.com/pmndrs/jotai/blob/09ac06216f537eae6d57cd344022e0182ae941bf/src/core/atom.ts#L23
async write means this `write` function returns a promise.

> Here's a code sandbox showing the solution: [async fetch to set primitive atoms via a write-only atom in jotai](https://codesandbox.io/s/async-fetch-to-set-primitive-atoms-via-a-write-only-atom-in-jotai-l1qdi?file=/src/View.tsx)

When I opened the csb, it doesn't show Loading...
Here's the modified version: https://codesandbox.io/s/async-fetch-to-set-primitive-atoms-via-a-write-only-atom-in-jotai-forked-jklyi?file=/src/atoms.ts:280-717
Is it what you expect?

**Comment by angusryer** [comment]
@dai-shi Thank you! That is what I was expecting.

- [x] Understand how to implement async atoms that trigger `Suspense` fallback

Recap for anyone who runs into this (correct me if I'm wrong):

**An atom will trigger the `Suspense` fallback only if:**

- the atom's write arg (2nd one) is async
- the awaited call is made directly and not from inside another containing function

This _will_ trigger the `Suspense` fallback:
``` typescript
const writeAtom = atom(null, async (get, set) => {
   const response = await new Promise<string>((resolve, _reject) => {
    setTimeout(() => {
      resolve("some returned value");
    }, 2000);
  });
  set(somePrimitiveAtom, "The returned value is: " + response);
})
```

This _will not_ trigger the `Suspense` fallback:
``` typescript
const writeAtom = atom(null, async (get, set) => {
   const getResponse = async () => {
      const response = await new Promise<string>((resolve, _reject) => {
        setTimeout(() => {
          resolve("some returned value");
        }, 2000);
      });
      set(somePrimitiveAtom, "The returned value is: " + response);
   }
   getResponse();
})
```

But **_both_** of the above will still properly set `somePrimitiveAtom` to the correct values.

Thanks again for your help!

**Comment by dai-shi** [comment]
@angusryer Thanks for the summary! Please do note this is about async `write`. In the readme, it's called "Async actions". We also have async `read`, which is "Derived async atoms" in the readme, and for this, all atoms having values (and their derived) will suspend.

**Comment by angusryer** [comment]
My pleasure, and thanks for the clarification on the terminology!

**Comment by dai-shi** [comment]
So, what's remaining is this?
> (add docs that only direct atom's suspend on write)

anyone can help?

**Comment by dai-shi** [comment]
#364 seems to be related.

**Comment by Noitidart** [comment]
Oh I'll hit that please, just in middle of catching up on the weeks work I got behind on a deadline this week. haha it happens from time to time

**Comment by Noitidart** [comment]
PR opened thanks to @angusryer's notes - https://github.com/JLarky/jotai/pull/1

**Comment by dai-shi** [comment]
Why pointing to JLarky repo??

**Comment by Noitidart** [comment]
Oh shoot haha. Will fix.

**Comment by Noitidart** [comment]
Fixed! hahah
https://github.com/pmndrs/website/pull/78
